src/state.rs

```diff
@@ -0,0 +1,178 @@
+    ) -> (
+        impl IntoIterator<Item = Apply<impl IntoIterator<Item = (H256, H256)>>>,
+        impl IntoIterator<Item = Log>,
+        impl IntoIterator<Item = PromiseCreateArgs>,
+    ) {
+        let (apply_iter, log_iter) = self.memory_stack_state.deconstruct();
+        (apply_iter, log_iter, self.promises.into_vec())
+    }
+}
+
+impl<'backend, 'config> AuroraState for AuroraStackState<'backend, 'config> {
+    fn add_promise(&mut self, promise: PromiseCreateArgs) {
+        self.promises.push(promise);
+    }
+}
+
+impl<'backend, 'config> Backend for AuroraStackState<'backend, 'config> {
+    fn gas_price(&self) -> U256 {
+        self.memory_stack_state.gas_price()
+    }
+
+    fn origin(&self) -> H160 {
+        self.memory_stack_state.origin()
+    }
+
+    fn block_hash(&self, number: U256) -> H256 {
+        self.memory_stack_state.block_hash(number)
+    }
+
+    fn block_number(&self) -> U256 {
+        self.memory_stack_state.block_number()
+    }
+
+    fn block_coinbase(&self) -> H160 {
+        self.memory_stack_state.block_coinbase()
+    }
+
+    fn block_timestamp(&self) -> U256 {
+        self.memory_stack_state.block_timestamp()
+    }
+
+    fn block_difficulty(&self) -> U256 {
+        self.memory_stack_state.block_difficulty()
+    }
+
+    fn block_gas_limit(&self) -> U256 {
+        self.memory_stack_state.block_gas_limit()
+    }
+
+    fn chain_id(&self) -> U256 {
+        self.memory_stack_state.chain_id()
+    }
+
+    fn exists(&self, address: H160) -> bool {
+        self.memory_stack_state.exists(address)
+    }
+
+    fn basic(&self, address: H160) -> Basic {
+        self.memory_stack_state.basic(address)
+    }
+
+    fn code(&self, address: H160) -> Vec<u8> {
+        self.memory_stack_state.code(address)
+    }
+
+    fn storage(&self, address: H160, index: H256) -> H256 {
+        self.memory_stack_state.storage(address, index)
+    }
+
+    fn original_storage(&self, address: H160, index: H256) -> Option<H256> {
+        self.memory_stack_state.original_storage(address, index)
+    }
+}
+
+impl<'backend, 'config> StackState<'config> for AuroraStackState<'backend, 'config> {
+    fn metadata(&self) -> &StackSubstateMetadata<'config> {
+        self.memory_stack_state.metadata()
+    }
+
+    fn metadata_mut(&mut self) -> &mut StackSubstateMetadata<'config> {
+        self.memory_stack_state.metadata_mut()
+    }
+
+    fn enter(&mut self, gas_limit: u64, is_static: bool) {
+        // TODO: What should we do with respect to gas limit
+        self.promises.enter();
+
+        self.memory_stack_state.enter(gas_limit, is_static);
+    }
+
+    fn exit_commit(&mut self) -> Result<(), ExitError> {
+        self.promises.commit();
+
+        self.memory_stack_state.exit_commit()
+    }
+
+    fn exit_revert(&mut self) -> Result<(), ExitError> {
+        self.promises.discard();
+
+        self.memory_stack_state.exit_revert()
+    }
+
+    fn exit_discard(&mut self) -> Result<(), ExitError> {
```

Comment:
Just curious, what is the difference between `exit_revert` and `exit_discard` in terms of semantics of the EVM?

Reply:
`exit_revert` is to signal a problem with the code on the given input (i.e. Runtime error triggered by an `assert/require`).

AFAIU `exit_discard` is called when some invariant about the EVM are violated. For example trying to [deploy two contracts to the same address](https://github.com/aurora-is-near/sputnikvm/blob/master/src/executor/stack/mod.rs#L387) or [nonce of some account is non zero](https://github.com/aurora-is-near/sputnikvm/blob/master/src/executor/stack/mod.rs#L392) while deploying a contract.